### PR TITLE
feat: added metric tools for measuring latency and creating histograms

### PR DESCRIPTION
**Pull request by asmie**

### CODE REVIEW

File: src/tools/metrics/recorder.rs [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+        if let MetricVal::Gauge(val) = *self.0.get(&key).unwrap() {
+            val
+        } else {
+            0.0
+        }
+    }
+
+    pub fn get_histogram(&self, metric: &'static str) -> Option<Vec<f64>> {
+        let key = CompositeKey::new(MetricKind::Histogram, Key::from_name(metric));
+        if let MetricVal::Histogram(vals) = self.0.get(&key).unwrap() {
```
Comment (lwojtow) [inline src/tools/metrics/recorder.rs:38]: Do we have to do it this way? This unwrapping seems to be very error-prone.

File: src/tools/metrics/recorder.rs [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+    CompositeKey, MetricKind,
+};
+
+// This is a false positive, `CompositeKey` does not depend on any interior mutable
+// types for its hashing implementation. Therefore, it is safe to use in our context.
+#[allow(clippy::mutable_key_type)]
+pub struct Snapshot(HashMap<CompositeKey, MetricVal>);
+
+impl Snapshot {
+    pub fn get_counter(&self, metric: &'static str) -> u64 {
```
Comment (lwojtow) [inline src/tools/metrics/recorder.rs:18]: Can we pass `Key` as the second arg instead?

File: src/tools/metrics/recorder.rs [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+
+        histogram.record(1.0);
+        histogram.record(3.0);
+        histogram.record(5.0);
+        histogram.record(9.0);
+
+        let snapshot = metrics.take_snapshot();
+        let constructed_histogram = snapshot.construct_histogram(METRIC_NAME).unwrap();
+
+        assert!(constructed_histogram.entries() == 4);
```
Comment (lwojtow) [inline src/tools/metrics/recorder.rs:192]: Why not `assert_eq` here?

File: src/tools/metrics/tables.rs [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+}
+
+impl std::fmt::Display for RequestsTable {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.write_str(&fmt_table(Table::new(&self.rows)))
+    }
+}
+
+/// Formats `f64` with 2 decimal points.
+pub fn table_float_display(x: &f64) -> String {
```
Comment (lwojtow) [inline src/tools/metrics/tables.rs:82]: `f64` is `Copy`

**Review by lwojtow — APPROVED** [review]

**Review by Rqnsom — APPROVED** [review]
I haven't checked too much in details we have similar files in other repos

File: src/tools/metrics/recorder.rs (hunk 2 above)
Comment (Rqnsom) [reply src/tools/metrics/recorder.rs:18]: seems like most of the API uses the metric name (`&str`), maybe we can keep it for consistency reasons

from below:
```rust
    fn can_get_counter_value() {
        let metrics = TestMetrics::default();
        let counter = register_counter!(METRIC_NAME);

        counter.increment(COUNTER_INC);

        let snapshot = metrics.take_snapshot();

        assert_eq!(snapshot.get_counter(METRIC_NAME), COUNTER_INC);
```

File: src/tools/metrics/tables.rs (hunk 4 above)
Comment (asmie) [reply src/tools/metrics/tables.rs:82]: Yes, but it's used with tabled to format data in the cells:
```
#[tabled(display_with = "table_float_display")]
    completion: f64,
```
Tabled demands there to be reference.

File: src/tools/metrics/recorder.rs (hunk 3 above)
Comment (asmie) [reply src/tools/metrics/recorder.rs:192]: Done.

**Pull request merged**